community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTrackingTest.java

```diff
@@ -0,0 +1,179 @@
+
+        // when
+        transactionIdTracking.assertUpToDate();
+
+        // then all good!
+    }
+
+    @Test( timeout = 500 )
+    public void shouldReturnIfTheVersionIsUpToDate() throws Exception
+    {
+        // given
+        long version = 5L;
+        when( transactionIdStore.getLastClosedTransactionId() ).thenReturn( version );
+        TransactionIdTracking transactionIdTracking =
+                new TransactionIdTracking( () -> transactionIdStore, version, 5, SECONDS );
+
+        // when
+        transactionIdTracking.assertUpToDate();
+
+        // then all good!
+
+    }
+
+    @Test( timeout = 300 )
+    public void shouldTimeoutIfTheVersionIsTooHigh() throws Exception
+    {
+        // given
+        long version = 5L;
+        when( transactionIdStore.getLastClosedTransactionId() ).thenReturn( version );
+        TransactionIdTracking transactionIdTracking =
+                new TransactionIdTracking( () -> transactionIdStore, version + 1, 100, MILLISECONDS );
+
+        // when
+        try
+        {
+            transactionIdTracking.assertUpToDate();
+            fail( "should have thrown" );
+        }
+        catch ( TransactionFailureException ex )
+        {
+            // then all good!
+        }
+    }
+
+    @Test( timeout = 2000 )
+    public void shouldBeKeepCheckingForNewVersionUntilTheTimeoutIsReached() throws Exception
+    {
+        // given
+        long version = 5L;
+        when( transactionIdStore.getLastClosedTransactionId() )
+                .thenReturn( version, version, version, version, version + 1 );
+        TransactionIdTracking transactionIdTracking =
+                new TransactionIdTracking( () -> transactionIdStore, version + 1, 5, SECONDS );
+
+        // when
+        transactionIdTracking.assertUpToDate();
+
+        // then all good!
+    }
+
+    @Test( timeout = 500 )
+    public void shouldBeAbleToUpdateTheVersion() throws Exception
+    {
+        // given
+        when( transactionIdStore.getLastClosedTransactionId() ).thenReturn( 42L );
+        long newVersion = 45L;
+        TransactionIdTracking transactionIdTracking =
+                new TransactionIdTracking( () -> transactionIdStore, newVersion, 10, MILLISECONDS );
+        try
+        {
+            transactionIdTracking.assertUpToDate();
+            fail( "should have thrown" );
+        }
+        catch ( TransactionFailureException e )
+        {
+            // expected
+        }
+
+        when( transactionIdStore.getLastClosedTransactionId() ).thenReturn( newVersion );
+
+        // when
+        transactionIdTracking.updateVersion( 46L );
+
+        try
+        {
+            transactionIdTracking.assertUpToDate();
+            fail( "should have thrown" );
+        }
+        catch ( TransactionFailureException e )
+        {
+            // then all good!
+        }
+    }
+
+    @Test( timeout = 500 )
+    public void shouldNotUpdateVersionIfTheInitialVersionIsLessThanZero() throws Exception
```

Comment:
Could the name include the why (no transactions yet applied) rather than the how (negative tx id)?